corpus_studies/solidity_state_study/slither_plugins/hasstate.py

```diff
@@ -0,0 +1,153 @@
+# and traverses it to find all of the variables used in the check.
+# It converts the variables to strings and returns them in a set.
+def get_vars_used(exp) :
+    if isinstance(exp, Identifier) :
+        return {str(exp)}
+    elif isinstance(exp, CallExpression) :
+        vars = [get_vars_used(arg) for arg in exp.arguments]
+        return reduce(lambda s,x : s.union(x), vars, set())
+    elif isinstance(exp, UnaryOperation) :
+        return get_vars_used(exp.expression)
+    elif isinstance(exp, BinaryOperation) :
+        return (get_vars_used(exp.expression_left)
+                    .union(get_vars_used(exp.expression_right))
+                )
+    elif isinstance(exp, Literal) :
+        return set()
+    elif isinstance(exp, IndexAccess) :
+        return (get_vars_used(exp.expression_left)
+                    .union(get_vars_used(exp.expression_right))
+                )
+    elif isinstance(exp, MemberAccess) :
+        return {str(exp)}
+    elif isinstance(exp, TupleExpression) :
+        # There should be exactly one element in the tuple 
+        #     (that is, they should act as parenthesis).
+        # Comparing tuples for equality as a Solidity language feature
+        #     does not exist at the time of writing.
+        return get_vars_used(exp.expressions[0])
+    return set()
+
+# Class implemented as a detector plugin for Slither. This detector detects,
+# for the given contracts, which contracts have state checks and 
+# thus are stateful
+class HasState(AbstractDetector):
+
+    STATEFUL_MESSAGE = "%s is stateful\n"
+
+    # Variables declared for use in Slither
+    ARGUMENT = 'hasstate'
+    HELP = 'Help printed by slither'
+    IMPACT = DetectorClassification.HIGH
+    CONFIDENCE = DetectorClassification.HIGH
+
+    WIKI = 'STATE TEST'
+
+    WIKI_TITLE = 'TODO'
+    WIKI_DESCRIPTION = 'Detects whether or not a contract is stateful'
+    WIKI_EXPLOIT_SCENARIO = '''
+    contract Pausable {
+        event Pause();
+        event Unpause();
+
+        bool public paused = false;
+
+        modifier whenNotPaused() {
+            require(!paused);
+            _;
+        }
+
+        modifier whenPaused {
+            require(paused);
+            _;
+        }
+
+        function pause() whenNotPaused returns (bool) {
+            paused = true;
+            Pause();
+            return true;
+        }
+
+        function unpause() whenPaused returns (bool) {
+            paused = false;
+            Unpause();
+            return true;
+        }
+    }
+    
+    In the above contract, there are two states checks, 
+    "require(paused)" and "require(!paused)". This contract is 
+    therefore stateful, and the detector will print out:
+    %s
+    ''' % (STATEFUL_MESSAGE % "Pausable")
+
+    WIKI_RECOMMENDATION = 'This is just a check, it does not indicate an error'
+
+    # Checks if any contracts that contract inherits were already determined 
+    # to be stateful. Returns the truth value of that check.
```

Comment:
What if we haven't processed those contracts yet? Does Slither process in order of the inheritance hierarchy?